File: src/bin/0152-maximum-product-subarray.py
# @leetup=info id=152 lang=python3 slug=maximum-product-subarray



# @leetup=code
class Solution:
    def maxProduct(self, nums: list[int]) -> int:
        positive_max = 1
        negative_max = 1

        res = -(1 << 32)
        for n in nums:
            if n == 0:
                positive_max = 1
                negative_max = 1
                res = max(res, 0)
            else:
                positive_max, negative_max = (
                    max(positive_max * n, negative_max * n, n),
                    min(positive_max * n, negative_max * n, n),
                )
                res = max(res, positive_max)
        return res


# @leetup=code

File: src/bin/0152-maximum-product-subarray.cpp
// @leetup=info id=152 lang=cpp slug=maximum-product-subarray

#include <climits>
#include <iostream>

// @leetup=code

#include <vector>
using namespace std;

class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    int positive_best = 1;
    int negative_best = 1;
    int res = INT_MIN;
    for (int& n : nums) {
      if (n == 0) {
        positive_best = 1;
        negative_best = 1;
        res = max(res, 0);
      } else {
        int p_temp = max(max(positive_best * n, negative_best * n), n);
        negative_best = min(min(positive_best * n, negative_best * n), n);
        positive_best = p_temp;
        res = max(res, positive_best);
      }
    }
    return res;
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}
